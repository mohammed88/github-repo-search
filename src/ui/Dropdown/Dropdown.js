import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import {lightGray, darkGray, orange} from '../styles/colors';
import Spacing from '../styles/spacing';


const Wrapper = styled.div`
	width: auto;
	position: relative;
	box-sizing: border-box;
`;
const Header = styled.div`
	border: 2px solid ${lightGray};
	border-radius: 3px;
	padding: ${Spacing.small} ${Spacing.medium};
`;
const HeaderTitle = styled.div`
	display: flex;
	justify-content: space-between;
`;
const List = styled.ul`
	background: white;
	border-radius: 3px;
	border: 2px solid ${lightGray};
	box-sizing: border-box;
	list-style-type: none;
	margin: ${Spacing.small} 0;
	padding: 0;
	position: absolute;
	width: inherit;
	width:100%;
	z-index:1;
`;
const ListItem = styled.li`
	padding: ${Spacing.small} ${Spacing.medium};
	border: 2px ${lightGray};
	border-style: none none solid none;
	:hover {
		background: ${orange};
	}

	:last-child {
		border: none;
	}
`;

class Dropdown extends React.Component {
	static propTypes = {
		list: PropTypes.array.isRequired,
		onChange: PropTypes.func,
		initialValue: PropTypes.string,
	};

	static defaultProps = {
		onChange: () => {},
		initialValue: '',
	};

	constructor(props) {
		super(props);
		this.wrapper = React.createRef();
	}

	state = {
		isOpen: false,
		value: '',
	};

	toggleList = (event={
		nativeEvent: {stopImmediatePropagation: () => {}},
		stopPropagation: () => {},
	}) => {
		event && event.stopPropagation();
		event.nativeEvent && event.nativeEvent.stopImmediatePropagation && event.nativeEvent.stopImmediatePropagation();

		this.setState({isOpen: !this.state.isOpen});
	};

	close = event => {
		if (this.state.isOpen && !this.wrapper.current.contains(event.target)) {
			this.toggleList();		
		}
	};

	selectItem = (index) => {
		const value = this.props.list[index];

		this.setState(
			{value, isOpen: false}, 
			() => this.props.onChange(this.state.value),
		);
	};

	componentDidMount() {
		document.addEventListener('click', this.close);
	}

	componentWillMount() {
		document.removeEventListener('click', this.close);
	}

	render() {
		const {isOpen, value} = this.state;
		const {list, initialValue} = this.props;

		return (
			<Wrapper ref={this.wrapper} data-component-type="Dropdown">
				<Header onClick={this.toggleList}>
					<HeaderTitle>
						<span>
							{value.length === 0 
								? initialValue.length === 0 
									? "Select Option" 
									: initialValue 
								: value
							}
						</span> 
						<span>
							{isOpen ? <CaretUpSvg /> : <CaretDownSvg />}
						</span>
					</HeaderTitle>
				</Header>

				{
					isOpen && 
					<List>
						{list.map((item, index) => 
							<ListItem 
								key={index} 
								onClick={this.selectItem.bind(this, index)}
								data-component-type="DropdownOption"
							>
								{item}
							</ListItem>)
						}
					</List>
				}
			</Wrapper>
		);
	}
}

function CaretUpSvg() {
	return (
		<svg 
			width="16" 
			height="16" 
			viewBox="0 0 1792 1792" xmlns="http://www.w3.org/2000/svg"
		>
			<path 
				d="
					M1408 1216q0 26-19 45t-45 19h-896q-26 
					0-45-19t-19-45 19-45l448-448q19-19 45-19t45 
					19l448 448q19 19 19 45z"
				fill={darkGray}
			/>
		</svg>
	);
}

function CaretDownSvg() {
	return (
		<svg 
			width="16" 
			height="16" 
			viewBox="0 0 1792 1792" 
			xmlns="http://www.w3.org/2000/svg"
		>
			<path 
				d="
					M1408 704q0 26-19 45l-448 448q-19 
					19-45 19t-45-19l-448-448q-19-19-19-45t19-45 
					45-19h896q26 0 45 19t19 45z"
				fill={darkGray}
			/>
		</svg>
	);
}

export default Dropdown;