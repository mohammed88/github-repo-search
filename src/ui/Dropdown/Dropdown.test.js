import React from 'react';
import {shallow, mount} from 'enzyme';

import Dropdown from './Dropdown';

describe('Dropdown', () => {
	it('should render', () => {
		const wrapper = shallow(<Dropdown list={[]} />);

		expect(wrapper).toMatchSnapshot();
	});

	it('should open options list onClick ', () => {
		const wrapper = mount(<Dropdown list={['option1', 'option2']} />);

		expect(wrapper.state().isOpen).toEqual(false);
		
		wrapper.find('span').first().simulate('click');

		expect(wrapper.state().isOpen).toEqual(true);
	});

	it('should display selected option', () => {
		const options = ['option1', 'option2'];
		const wrapper = mount(<Dropdown list={options} />);

		// open dropdown
		wrapper.find('span').first().simulate('click');
		// select first option
		wrapper.find('li').first().simulate('click');

		expect(wrapper.state().isOpen).toEqual(false);
		expect(wrapper.state().value).toEqual('option1');
	});
});