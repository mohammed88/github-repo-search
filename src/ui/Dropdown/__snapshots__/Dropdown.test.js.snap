// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dropdown
    initialValue=""
    list={Array []}
    onChange={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <ForwardRef
          onClick={[Function]}
        >
          <ForwardRef>
            <span>
              Select Option
            </span>
            <span>
              <CaretDownSvg />
            </span>
          </ForwardRef>
        </ForwardRef>,
        false,
      ],
      "data-component-type": "Dropdown",
    },
    "ref": Object {
      "current": null,
    },
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <ForwardRef>
            <span>
              Select Option
            </span>
            <span>
              <CaretDownSvg />
            </span>
          </ForwardRef>,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <span>
                Select Option
              </span>,
              <span>
                <CaretDownSvg />
              </span>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Select Option",
              },
              "ref": null,
              "rendered": "Select Option",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <CaretDownSvg />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "span",
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-htpNat",
              "isStatic": true,
              "rules": Array [
                "
	display: flex;
	justify-content: space-between;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-htpNat",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bwzfXH",
            "isStatic": true,
            "rules": Array [
              "
	border: 2px solid ",
              "#EDEDED",
              ";
	border-radius: 3px;
	padding: ",
              "8px",
              " ",
              "12px",
              ";
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bwzfXH",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      false,
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-bdVaJa",
        "isStatic": true,
        "rules": Array [
          "
	width: auto;
	position: relative;
	box-sizing: border-box;
",
        ],
      },
      "displayName": "styled.div",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-bdVaJa",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ForwardRef
            onClick={[Function]}
          >
            <ForwardRef>
              <span>
                Select Option
              </span>
              <span>
                <CaretDownSvg />
              </span>
            </ForwardRef>
          </ForwardRef>,
          false,
        ],
        "data-component-type": "Dropdown",
      },
      "ref": Object {
        "current": null,
      },
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <ForwardRef>
              <span>
                Select Option
              </span>
              <span>
                <CaretDownSvg />
              </span>
            </ForwardRef>,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <span>
                  Select Option
                </span>,
                <span>
                  <CaretDownSvg />
                </span>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Select Option",
                },
                "ref": null,
                "rendered": "Select Option",
                "type": "span",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <CaretDownSvg />,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": "span",
              },
            ],
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-htpNat",
                "isStatic": true,
                "rules": Array [
                  "
	display: flex;
	justify-content: space-between;
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-htpNat",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            },
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bwzfXH",
              "isStatic": true,
              "rules": Array [
                "
	border: 2px solid ",
                "#EDEDED",
                ";
	border-radius: 3px;
	padding: ",
                "8px",
                " ",
                "12px",
                ";
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bwzfXH",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        false,
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bdVaJa",
          "isStatic": true,
          "rules": Array [
            "
	width: auto;
	position: relative;
	box-sizing: border-box;
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-bdVaJa",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
